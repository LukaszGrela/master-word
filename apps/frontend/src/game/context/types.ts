import { TGameRecord, TScore } from '@repo/backend-types';
import { IDictionary } from '@repo/common-types';
import { TErrorResponse } from '../../api';

export interface IGameContext {
  init: (
    language: string,
    maxAttempts: number,
    wordLength: number,
    session?: string,
  ) => void;
  guess: (word: string) => void;

  loading: boolean;
  error?: TErrorResponse;

  session?: string;

  language: string;
  maxAttempts: number;
  wordLength: number;

  highest?: IDictionary<TScore>;
  game?: TGameRecord;
}
