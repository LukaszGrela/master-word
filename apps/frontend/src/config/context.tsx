import React from 'react';

import { IConfigContext } from './types';

export const ConfigContext = React.createContext<IConfigContext>(null!);
