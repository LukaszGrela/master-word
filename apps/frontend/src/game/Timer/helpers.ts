import { toHMS } from '@repo/utils';

const prefix0 = (n: number): string => (n < 10 ? `0${n}` : `${n}`);

export const displayTime = (ms: number): string => {
  const { h, m, s } = toHMS(ms);

  if (!h && !m) {
    return `00:${prefix0(s >> 0)}`;
  }

  if (!h && m !== undefined) {
    return `${prefix0(m)}:${prefix0(s)}`;
  }

  if (h !== undefined && m !== undefined) {
    return `${prefix0(h)}:${prefix0(m)}:${prefix0(s)}`;
    /* v8 ignore start */
  }

  return '00:00'; // it should never reach here
};
/* v8 ignore stop */
