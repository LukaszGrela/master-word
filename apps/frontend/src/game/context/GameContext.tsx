import React from 'react';
import { IGameContext } from './types';

export const GameContext = React.createContext<IGameContext>(null!);
