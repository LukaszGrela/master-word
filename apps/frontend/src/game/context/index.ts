/* v8 ignore start */
export * from './types';
export * from './GameContext';
export * from './GameProvider';
export * from './hooks';
/* v8 ignore stop */
