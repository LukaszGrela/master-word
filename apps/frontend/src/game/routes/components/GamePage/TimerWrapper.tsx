import { FC } from 'react';
import { Timer } from '../../../Timer';
import { useGameContext } from '../../../context';

export const TimerWrapper: FC = () => {
  const { loading, game } = useGameContext();
  return (
    <Timer
      startMs={
        !loading && game && game.attempt > 0 && !game.finished
          ? Number(game.timestamp_start)
          : undefined
      }
    />
  );
};
