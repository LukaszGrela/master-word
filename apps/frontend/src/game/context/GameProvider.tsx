import { FC, useCallback, useMemo, useRef, useState } from 'react';
import { useImmer } from 'use-immer';
import { TGameSession } from '@repo/backend-types';
import { GameContext } from './GameContext';
import { IGameContext } from './types';
import {
  TErrorResponse,
  getInit,
  getNextAttempt,
  toErrorResponse,
} from '../../api';
import { ATTEMPTS, WORD_LENGTH } from '../constants';

const SESSION_NOT_SET = 'session-not-set' as const;
export const GameProvider: FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<TErrorResponse | undefined>();

  const [gameData, setGameData] = useImmer<
    TGameSession & Pick<IGameContext, 'language' | 'maxAttempts' | 'wordLength'>
  >({
    session: SESSION_NOT_SET,
    language: 'pl',
    maxAttempts: ATTEMPTS,
    wordLength: WORD_LENGTH,
  });

  const initBusy = useRef<boolean>(false);
  const init = useCallback(
    (
      language: string,
      maxAttempts: number,
      wordLength: number,
      session?: string,
    ) => {
      if (initBusy.current === false) {
        initBusy.current = true;
        setError(undefined);
        setLoading(true);
        setGameData((draft) => {
          draft.language = language;
          draft.maxAttempts = maxAttempts;
          draft.wordLength = wordLength;
        });
        getInit({
          language,
          maxAttempts,
          session,
          wordLength,
        })
          .then((gameSession) => {
            setGameData((draft) => {
              draft.session = gameSession.session;
              draft.game = gameSession.game;
              if (gameSession.highest) {
                draft.highest = gameSession.highest;
              } else {
                draft.highest = undefined;
              }
              if (gameSession.game) {
                draft.wordLength = gameSession.game.word_length;
                draft.maxAttempts = gameSession.game.max_attempts;
                draft.language = gameSession.game.language;
              }
            });
            setLoading(false);
          })
          .catch((error: unknown) => {
            // abort controller errors are ok
            if (!(error instanceof DOMException)) {
              // others are logged
              setError(toErrorResponse(error as Error));
              setLoading(false);
            }
          })
          .finally(() => {
            initBusy.current = false;
          });
      }
    },
    [setGameData],
  );

  const guessBusy = useRef<boolean>(false);
  const guess = useCallback(
    (word: string) => {
      if (
        gameData.session !== SESSION_NOT_SET &&
        guessBusy.current === false &&
        initBusy.current === false
      ) {
        guessBusy.current = true;
        setError(undefined);

        setLoading(true);
        getNextAttempt({
          guess: word.toLocaleUpperCase(),
          session: gameData.session,
        })
          .then((gameState) => {
            setGameData((draft) => {
              if (!draft.game) {
                draft.game = gameState.game;
              } else {
                if (gameState.game.finished) {
                  draft.highest = gameState.highest;
                  draft.game = gameState.game;
                } else {
                  const { attempt, state } = gameState.game;
                  if (attempt === 1) {
                    draft.game.timestamp_start = gameState.game.timestamp_start;
                  }
                  draft.game.attempt = attempt;
                  const row = draft.game.state.at(attempt - 1);
                  const sourceRow = state.at(attempt - 1);
                  if (row && sourceRow) {
                    // update only current state
                    row.word = sourceRow.word;
                    row.validated = sourceRow.validated;
                  }
                }
              }
            });
            setLoading(false);
          })
          .catch((error: unknown) => {
            // abort controller errors are ok
            if (!(error instanceof DOMException)) {
              // others are logged
              const enrichedError = toErrorResponse(error as Error);
              if (enrichedError.code === 6 /* INVALID_WORD */) {
                enrichedError.details = word;
              }
              setError(enrichedError);
              setLoading(false);
            }
          })
          .finally(() => {
            guessBusy.current = false;
          });
      }
    },
    [gameData.session, setGameData],
  );

  const value = useMemo((): IGameContext => {
    return {
      loading,
      init,
      guess,

      error,

      language: gameData.language,
      maxAttempts: gameData.maxAttempts,
      wordLength: gameData.wordLength,
      game: gameData?.game,
      highest: gameData?.highest,
      session:
        gameData.session === SESSION_NOT_SET ? undefined : gameData.session,
    };
  }, [
    error,
    gameData?.game,
    gameData?.highest,
    gameData.language,
    gameData.maxAttempts,
    gameData.session,
    gameData.wordLength,
    guess,
    init,
    loading,
  ]);
  return <GameContext.Provider value={value}>{children}</GameContext.Provider>;
};
