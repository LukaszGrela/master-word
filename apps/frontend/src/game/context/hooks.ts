import React from 'react';
import { GameContext } from './GameContext';

export function useGameContext() {
  return React.useContext(GameContext);
}
