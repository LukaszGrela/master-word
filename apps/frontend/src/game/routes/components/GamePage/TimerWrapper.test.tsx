import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { act } from '@testing-library/react';
import {
  contextRenderer,
  gameResponses,
} from '../../../context/__tests__/contextRenderer';
import { IGameContext } from '../../../context';
import { TimerWrapper } from './TimerWrapper';
import { createFetchResponse } from '../../../../__tests__/helpers';
import { TGameRecord } from '@repo/backend-types';

const originalFetch = global.fetch;

describe('game', () => {
  describe('routes', () => {
    describe('components', () => {
      describe('GamePage', () => {
        describe('TimerWrapper', () => {
          let mockFetch = vi.fn();

          beforeEach(() => {
            vi.useFakeTimers();
            vi.setSystemTime(new Date(1979, 5, 13));
            mockFetch = vi.fn();
            global.fetch = mockFetch;

            // MOCK FETCH FIRST
            mockFetch.mockResolvedValue(
              createFetchResponse(true, gameResponses.gameInit),
            );
          });
          afterEach(() => {
            vi.clearAllMocks();
            global.fetch = originalFetch;
            vi.useRealTimers();
          });

          it('renders', () => {
            const { container } = contextRenderer(<TimerWrapper />);

            expect(container).toMatchSnapshot();
          });

          it("doesn't change time while loading", () => {
            const gameContext: Pick<IGameContext, 'loading'> = {
              loading: false,
            };
            const { getByText } = contextRenderer(<TimerWrapper />, {
              context: gameContext,
            });

            expect(getByText('00:00')).toBeDefined();

            // progress time
            act(() => {
              vi.advanceTimersByTime(3000);
            });

            // it still need to say 00:00
            expect(getByText('00:00')).toBeDefined();
          });

          it("doesn't change time when game attempt is 0", () => {
            const gameContext: Pick<IGameContext, 'loading' | 'game'> = {
              loading: false,
              game: gameResponses.gameInit.game as unknown as TGameRecord,
            };
            const { getByText } = contextRenderer(<TimerWrapper />, {
              context: gameContext,
            });

            expect(getByText('00:00')).toBeDefined();

            // progress time
            act(() => {
              vi.advanceTimersByTime(3000);
            });

            // it still need to say 00:00
            expect(getByText('00:00')).toBeDefined();
          });

          it("doesn't change time when game is finished", () => {
            const gameContext: Pick<IGameContext, 'loading' | 'game'> = {
              loading: false,
              game: gameResponses.gameWon.game as unknown as TGameRecord,
            };
            const { getByText } = contextRenderer(<TimerWrapper />, {
              context: gameContext,
            });

            expect(getByText('00:00')).toBeDefined();

            // progress time
            act(() => {
              vi.advanceTimersByTime(3000);
            });

            // it still need to say 00:00
            expect(getByText('00:00')).toBeDefined();
          });

          it('Updates timer while game is running', () => {
            const gameContext: Pick<IGameContext, 'loading' | 'game'> = {
              loading: false,
              game: {
                ...gameResponses.gameNext.game,
                timestamp_start: new Date(1979, 5, 12, 23, 1, 1)
              } as unknown as TGameRecord,
            };
            const { getByText } = contextRenderer(<TimerWrapper />, {
              context: gameContext,
            });

            expect(getByText('00:00')).toBeDefined();

            // progress time
            act(() => {
              vi.advanceTimersByTime(3000);
            });

            expect(getByText('59:02')).toBeDefined();
          });
        });
      });
    });
  });
});
