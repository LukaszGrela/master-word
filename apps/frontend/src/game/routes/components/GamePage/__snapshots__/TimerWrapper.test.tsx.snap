// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > routes > components > GamePage > TimerWrapper > renders 1`] = `
<div>
  <span
    class="timer"
  >
    00:00
  </span>
</div>
`;
